### DISCUSSION TITLE
Unable to activate ST7789 using STM32F411 - SOLVED

### DISCUSSION THREAD

**Discussion by nickhileu** [Bodmer/TFT_eSPI#1678] [General]
I have spent the last 2 days trying to make this library work using above combination.
Board: STM32F411CEU6 , ST official core
Screen: ST7789 TFT 135x240 

This combination is working with Adafruit ST7789 library however I am suffering performance issue, thus I am very keen to make this work.

What I have tried:
1. tried both ST7789_DRIVER and ST7789_2_DRIVER (of course one at a time)
2. tried Read_user_setup but I have to nothing to the output, I have ruled out serial print issue (printed some dummy message to verify)
3. tried Setup_135_St7789 and User_setup.h with proper includes from user_setup_select.h
4. tried both PIN number and integer number (I don't think it has any difference anyways, both works in Adafruit)5. 

Below is my setup file, what am I doing wrong here?

Thanks in advance.
Nick

`// ST7789 135 x 240 display with no chip select line

#define ST7789_DRIVER     // Configure all registers

#define STM32

#define TFT_WIDTH  135
#define TFT_HEIGHT 240

//#define CGRAM_OFFSET      // Library will add offsets required

//#define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

#define TFT_INVERSION_ON
//#define TFT_INVERSION_OFF

// DSTIKE stepup
#define TFT_DC   20//PB4 
#define TFT_RST   22//PB6
#define TFT_MOSI  21//PB5
#define TFT_SCLK  19//PB3
#define TFT_CS  8 // PA8 // Not connected
#define TFT_BL   23//PB7            // LED back-light control pin PB7
#define TFT_BACKLIGHT_ON HIGH  // Level to turn ON back-light (HIGH or LOW)

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

// #define SPI_FREQUENCY  27000000
#define SPI_FREQUENCY  40000000

//#define SPI_READ_FREQUENCY  20000000

//#define SPI_TOUCH_FREQUENCY  2500000

 #define SUPPORT_TRANSACTIONS
`

**Comment by Bodmer** [comment]
You must use the hardware SPI port pins because TFT_eSPI does not support software SPI.

The Adafruit library will "bit bash" SPI using software, thus will be slow.

Correct pin allocations from data sheet:
![20220302_130801](https://user-images.githubusercontent.com/15824805/156368206-2a94d323-5199-41f1-b074-3ad3d562c7a7.jpg)

**Comment by Bodmer** [comment]
To use port B:
![Screenshot_20220302-131614_Drive](https://user-images.githubusercontent.com/15824805/156369047-5a9c8661-5328-41c9-aafd-89350268e502.jpg)

**Comment by nickhileu** [comment]
Thanks for the reply. Yes, you are right, I was using software SPI on Adafruit , which I did not realise, because I was using hardware SPI pins. 
For the record, what made your library working with ST official core was I had to define 
 `#define TFT_MISO  PA6  `
even the screen does not use MISO
It is not your fault and Adafruit lib also requires this to be set to enable hardware SPI.


Now I can play with the demos, I currently would like to load some fonts and do flicker free refresh on text changed, thus I am trying to play with font demo, however have below error.  
`Smooth_font_gradient:79:9: error: 'class TFT_eSPI' has no member named 'loadFont'
   79 |     tft.loadFont(AA_FONT_LARGE);
      |         ^~~~~~~~`

Which demo would be a good starting point for font / text refresh ?

Thanks 
Nick

**Reply by Bodmer** [reply]
It sounds like #define SMOOTH_FONT is not present in the setup file. When it is defined then loadFont should be available, and the examples should run.

You can run the diagnostic sketch "Read_User_Setup" to see the settings picked up by the compiler in the Serial Monitor widow.
(Pins number might show as garbage on an STM32, havent figured out how to handle that yet)

Smooth fonts sonsume large amounts of FLASH and RAM...

**Reply by Bodmer** [reply]
Use the "Array" examples, as the STM32 package does not have SPIFFS or LittleFS.

**Reply by nickhileu** [reply]
I tried SMOOTH_FONT and yes, it uses quite FLASH and RAM.
How would I load Adafruit fonts if I have defined LOAD_GFXFF ?

I have already seen such big performance improvement vs Adafruit lib! It looks very promising!

**Reply by Bodmer** [reply]
I call the Adafruit GFXFF fonts the "Free fonts" so there are some examples:
```
All_Free_Fonts_Demo
Free_Font_Demo
TFT_Custom_Fonts
```

**Comment by nickhileu** [comment]
Excellent! This library is fantastic! I have put in to my code and its behaviour is more predictable than Adafruit.
Two questions though:
1. what happens if I want to support multiple screens with same controller IC but different resolution in one project? For example user flick some switches to choose which one to use, which means my code dynamically switch concrete screen implementation. Currently it is in user_setup_select header file, but only one setup can be loaded. How to proceed?
2. I don't quite get `TFT_eSprite` behaviour, say if I print a long string "abcd" and print a shorter one "abc" , it does not mask off "d" even I do a `sprite fillScreen(colour)` before print. Is this expected?

Cheers,
Nick

**Comment by Bodmer** [comment]
Thanks for the feedback, it's good to hear it is works. Re. questions.

1. The library is not coded to permit multiple instantiations with different settings. However it is possible to use 2 screens of the same type if the two chip selects are controlled by the sketch. In this case, do not define TFT_CS in the library setup file,  select the TFT to talk to by setting only one chip select low in the sketch. Only set both chip selects low at the same time during tft.init(). Configure the setup file for the largest display. If the second screen has fewer pixels you should use the viewport functions (see examples) to restrict graphics to the display area on the smaller screen. You will need to experiment to determine the viewport offsets and this will change with screen rotation.
2. This is as expected. A sprite is like a small screen that can be pasted onto the real screen. Fill with a different colour as a test and you will see what I mean. Run examplles, reading sketch header and comments should help.

Work with one screen initially to familiariise yourself with the library. There are lots of examples that cover the library functions and most have explanation comments

**Comment by nickhileu** [comment]
What I didn't mention was I declared sprite size exactly as screen size. 
I think what was confusing was the behaviour of sprite on a rotated screen.
For example:
If I declared a screen W 135 H 240
then set rotation to 1 
and declared a sprite size of w 240 h 135
and do a sprite fullscreen , push it to (0,0)
left part of screen is filled, rather than expected full screen is filled.

I have tried different combination of swapping W and H of screen and sprite as well as set rotation on screen , with setting different offset on `sprite push`  but only combination worked as I thought it should was they both at same W and H and screen rotation is 0.

How to deal with sprite on a landscape screen?

Cheers,
Nick

**Reply by Bodmer** [reply]
The ST7789 is a 240x320 pixel driver. So the location of a lower resolution screen can be anywhere in this area and in any rotation the manufacturer decided upon.

An analogy is: take a piece of paper, draw a rectangle in the top left corner (i.e at 0,0), assume that is the area the low res screen occupies within the 240x320 area, now rotate the paper by 90 or 180 deg. Now if you draw a rectangle at the top left (0,0) it is not going to fully overlap the the visible area.

You need to find the offsets to the visible area in the total 240x320 drivable area. You can do this be experimentation.  Normally the library compensates for this with predefined offsets for common screens, see here:
https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_Drivers/ST7789_Rotation.h

For each rotation you will see lines like this:
```
      if (_init_width == 135)
      {
        colstart = 52;
        rowstart = 40;
      }
```
These offsets change with the rotations. If you look you will see offsets of 52 and 53 are used for example, this is because 52+135+53 = 240. So the manufaturer has put the 135 pixel visible screen area near the middle of the 240 wide driveable area.

Since you have two screens of different resolutions then set the library to the highest res and compensate in your sketch for the offsets in the rotations you decide to use.
